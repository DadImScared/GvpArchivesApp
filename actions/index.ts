
import itemsByCategoryactions, { ItemsByCategoryActions } from "./itemsByCategory";
import itemsByIdActions, { ItemsByIdActions } from "./itemsById";
import loadingActions, { ILoadingActions } from "./loading";

interface IActions {
  loading: ILoadingActions;
  itemsByCategory: ItemsByCategoryActions;
  itemsById: ItemsByIdActions;
}

const actions: IActions = {
  itemsByCategory: itemsByCategoryactions,
  itemsById: itemsByIdActions,
  loading: loadingActions
};

export default actions;
