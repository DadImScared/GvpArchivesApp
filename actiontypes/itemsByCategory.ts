export const GET_ITEMS_BY_CATEGORY = "itemsByCategory/GET_ITEMS_BY_CATEGORY";
export const ADD_ITEMS_BY_CATEGORY = "itemsByCategory/ADD_ITEMS_BY_CATEGORY";
export const UPDATE_ITEMS_BY_CATEGORY = "itemsByCategory/UPDATE_ITEMS";
