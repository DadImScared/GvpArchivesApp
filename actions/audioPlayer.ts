import { action } from "typesafe-actions";

import * as AudioActionTypes from "../actiontypes/audioPlayer";

export const togglePlaying = () => action(AudioActionTypes.TOGGLE_PLAYING);

export const setPlaying = (playing: boolean) => action(AudioActionTypes.SET_PLAYING, { playing });

export const showPlayer = (shouldShowPlayer: boolean) => {
  return action(AudioActionTypes.SHOW_PLAYER, { showPlayer: shouldShowPlayer });
};

export const setPlaylist = (name: string, items: string[], currentIndex = 0) => {
  return action(
    AudioActionTypes.SET_PLAYLIST,
    {
      currentIndex,
      items,
      name,
      playerType: "playlist" as any,
      playing: true,
      showPlayer: true,
    }
  );
};

export const updateIndex = (newIndex: number) => action(AudioActionTypes.UPDATE_CURRENT_INDEX, { newIndex });

export const setSong = (songUrl: string, songName: string) => {
  return action(
    AudioActionTypes.SET_SONG,
    {
      songName,
      songUrl,
      playerType: "song",
      playing: true,
      showPlayer: true
    }
  );
};

export interface IAudioPlayerActions {
  updateIndex: typeof updateIndex;
  togglePlaying: typeof togglePlaying;
  setSong: typeof setSong;
  setPlaying: typeof setPlaying;
  setPlaylist: typeof setPlaylist;
  showPlayer: typeof showPlayer;
}

export default {
  setPlaying,
  setPlaylist,
  setSong,
  showPlayer,
  togglePlaying,
  updateIndex,
};
