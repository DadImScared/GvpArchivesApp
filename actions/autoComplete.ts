import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/autoComplete";

const addAutoComplete = (query: string, categories: string[], results: string[]) => {
  return action(ActionTypes.ADD_AUTOCOMPLETE, { query, categories, results });
};

export default {
  addAutoComplete
};
