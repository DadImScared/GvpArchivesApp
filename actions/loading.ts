import { action } from "typesafe-actions";

import { LOADING_END, LOADING_START } from "../actiontypes/loading";

export const loadingStart = (id: string) => action(LOADING_START, id);

export const loadingEnd = (id: string) => action(LOADING_END, id);

export interface ILoadingActions {
  loadingStart: typeof loadingStart;
  loadingEnd: typeof loadingEnd;
}

export default {
  loadingEnd,
  loadingStart
};
