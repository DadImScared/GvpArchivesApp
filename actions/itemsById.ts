import { action } from "typesafe-actions";
import { ADD_ITEMS } from "../actiontypes/itemsById";
import { Item } from "../reducers/itemsById";

export const BASE_URL = __DEV__ ? "http://192.168.0.4:3000/api/v1" : "https://api.gvparchives.com/api/v1";

export const addItemsById = (items: { [key: string]: Item }) => action(ADD_ITEMS, { items });

export interface ItemsByIdActions {
  addItemsById: typeof addItemsById;
}

export default {
  addItemsById
};
