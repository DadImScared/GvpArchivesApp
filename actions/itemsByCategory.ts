import axios from "axios";
import { normalize } from "normalizr";
import { Action, ActionCreator } from "redux";
import { batchActions } from "redux-batched-actions";
import { ThunkAction } from "redux-thunk";
import { action } from "typesafe-actions";
import { ADD_ITEMS_BY_CATEGORY, UPDATE_ITEMS_BY_CATEGORY } from "../actiontypes/itemsByCategory";
import { ITEMS_BY_CATEGORY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { itemSchema } from "../schemas/items";
import { addItemsById, BASE_URL } from "./itemsById";
import { loadingEnd, loadingStart } from "./loading";

export type ActionCreatorThunk = ActionCreator<ThunkAction<void, IReducerState, undefined, Action<any>>>;

export const getItemsByCategory: ActionCreatorThunk = (category: string, nextPage?: string | boolean) => {
  return async (dispatch): Promise<void> => {
    const LOADING_ID = `${ITEMS_BY_CATEGORY}${category}`;
    dispatch(loadingStart(LOADING_ID));
    const firstPage = `${BASE_URL}/api/v1/items/${category}`;
    const url = typeof nextPage === "string" ? `${BASE_URL}${nextPage}` : firstPage;
    const response = await axios.get(url);
    if (response) {
      // dispatch batch actions
      const { data } = response;
      const { result: itemIds, entities: { items } } = normalize(data.results, itemSchema);
      dispatch(batchActions([
        addItemsById(items),
        (nextPage ? updateItemsByCategory : addItemsByCategory)({category, itemIds, nextPage: data.nextPage }),
        loadingEnd(LOADING_ID)
      ]));
    }
  };
};

interface ItemsByCategory {
  category: string;
  itemIds: string[];
  nextPage: boolean | string;
}

export const addItemsByCategory  = (data: ItemsByCategory) => {
  return action(ADD_ITEMS_BY_CATEGORY, data);
};

export const updateItemsByCategory = (data: ItemsByCategory) => {
  return action(UPDATE_ITEMS_BY_CATEGORY, data);
};

export interface ItemsByCategoryActions {
  addItemsByCategory: typeof addItemsByCategory;
  updateItemsByCategory: typeof updateItemsByCategory;
}

export default {
  addItemsByCategory,
  updateItemsByCategory
};
