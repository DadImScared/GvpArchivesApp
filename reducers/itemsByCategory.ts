import _ from "lodash";
import moment from "moment";
import { NavigationScreenProps } from "react-navigation";
import { createSelector } from "reselect";
import { ActionType } from "typesafe-actions";

import actions from "../actions";
import { ADD_ITEMS_BY_CATEGORY, UPDATE_ITEMS_BY_CATEGORY } from "../actiontypes/itemsByCategory";
import { ITEMS_BY_CATEGORY } from "../loadingPrefix";
import { IReducerState } from "./index";
import { getItemsById, Item } from "./itemsById";
import { getLoading } from "./loading";

interface ItemsByCategory {
  results: string[];
  nextPage?: boolean | string;
  timeToLive?: number;
}

export interface ItemsByCategoryData {
  results: Item[];
  isLoading: boolean;
  nextPage: boolean | string;
  timeToLive: number;
}

export interface ItemsByCategoryState {
  [key: string]: ItemsByCategory;
}

function itemsByCategory(
    state: ItemsByCategoryState = {},
    action: ActionType<typeof actions.itemsByCategory>
) {
  switch (action.type) {
    case ADD_ITEMS_BY_CATEGORY:
      return {
        ...state,
        [action.payload.category]: {
          nextPage: action.payload.nextPage,
          results: action.payload.itemIds,
          timeToLive: moment().add(1, "days").valueOf()
        }
      };
    case UPDATE_ITEMS_BY_CATEGORY:
      return {
        ...state,
        [action.payload.category]: {
          ...state[action.payload.category],
          nextPage: action.payload.nextPage,
          results: _.union(state[action.payload.category].results, action.payload.itemIds)
        }
      };
    default:
      return state;
  }
}

export const getItems = (state: IReducerState) => state.itemsByCategory;

export const getCategory = (_state: any, props: NavigationScreenProps) => props.navigation.getParam("category");

export const getItemsByCategoryState = createSelector(
    [getItems, getCategory],
    (items, category) => items[category] || {}
);

export const getItemsByCategory = () => {
  return createSelector(
      [
        getItemsByCategoryState,
        getItemsById,
        getLoading(ITEMS_BY_CATEGORY, "navigation.state.params.category")
      ],
      ({ results = [], nextPage = false, timeToLive = 0 }, items, isLoading): ItemsByCategoryData => ({
        isLoading,
        nextPage,
        timeToLive,
        results: results.map((itemId) => items[itemId])
      })
  );
};

export default itemsByCategory;
