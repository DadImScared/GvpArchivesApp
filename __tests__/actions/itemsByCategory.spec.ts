import axios from "axios";
import { addItemsByCategory, getItemsByCategory, updateItemsByCategory } from "../../actions/itemsByCategory";

jest.mock("axios");
const payload = { category: "book", itemIds: ["1", "2"], nextPage: false};

describe("itemsByCategory", () => {
  test("addItemsByCategory", () => {
    expect(addItemsByCategory(payload)).toMatchSnapshot();
  });

  test("updateItemsbyCategory", () => {
    expect(updateItemsByCategory(payload)).toMatchSnapshot();
  });

  describe("getItemsByCategory", () => {
    it("should dispatch two times if not currently loading", async () => {
      (axios.get as any).mockImplementationOnce(() => {
        return Promise.resolve({
          data: {
            nextPage: false,
            results: []
          }
        });
      });
      const dispatch = jest.fn();
      const getState = () => ({
        itemsByCategory: {},
        itemsById: {},
        loading: {
          ITEMS_BY_CATEGORY_book: false
        },
      });
      await getItemsByCategory("book", false)(dispatch, getState, undefined);
      expect(dispatch).toHaveBeenCalledTimes(2);
    });
  });
});
