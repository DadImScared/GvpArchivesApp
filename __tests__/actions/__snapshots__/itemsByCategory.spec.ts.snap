// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`itemsByCategory addItemsByCategory 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "category": "book",
    "itemIds": Array [
      "1",
      "2",
    ],
    "nextPage": false,
  },
  "type": "itemsByCategory/ADD_ITEMS_BY_CATEGORY",
}
`;

exports[`itemsByCategory updateItemsbyCategory 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "category": "book",
    "itemIds": Array [
      "1",
      "2",
    ],
    "nextPage": false,
  },
  "type": "itemsByCategory/UPDATE_ITEMS",
}
`;
